### ISSUE TITLE
remove_configuration_by_configuration() may remove multiple configurations

### ISSUE THREAD

**Issue by juhaerk** [activeldap/activeldap#155]
In lib/active_ldap/configuration.rb remove_configuration_by_configuration() may remove multiple configurations.  It is called from remove_connection() in lib/active_ldap/connection.rb with configuration value, but unfortunately if there is another configuration with the same values, but associated with a different key, that configuration gets removed as well!

One fix is to check the key as well and remove only the configuration associated with that key. Here's a patch to do that... (not tested much, yet, but sending anyway).

```
diff -ruN activeldap-5.1.0.orig/lib/active_ldap/configuration.rb activeldap-5.1.0/lib/active_ldap/configuration.rb
--- activeldap-5.1.0.orig/lib/active_ldap/configuration.rb      2018-06-04 12:27:46.895076647 +0300
+++ activeldap-5.1.0/lib/active_ldap/configuration.rb   2018-06-04 13:06:28.219333185 +0300
@@ -91,8 +91,10 @@
         @@defined_configurations
       end
 
-      def remove_configuration_by_configuration(config)
-        @@defined_configurations.delete_if {|key, value| value == config}
+      def remove_configuration_by_key_and_configuration(key, config)
+        @@defined_configurations.delete_if do |_key, _config|
+          _key == key && _config == config
+        end
       end
 
       CONNECTION_CONFIGURATION_KEYS = [:uri, :base, :adapter]
diff -ruN activeldap-5.1.0.orig/lib/active_ldap/connection.rb activeldap-5.1.0/lib/active_ldap/connection.rb
--- activeldap-5.1.0.orig/lib/active_ldap/connection.rb 2018-06-04 12:27:46.895076647 +0300
+++ activeldap-5.1.0/lib/active_ldap/connection.rb      2018-06-04 13:06:51.091730183 +0300
@@ -135,8 +135,10 @@
         end
         config = configuration(key)
         conn = active_connections[key]
-        remove_configuration_by_configuration(config)
-        active_connections.delete_if {|_key, value| value == conn}
+        remove_configuration_by_key_and_configuration(key, config)
+        active_connections.delete_if do |_key, _config|
+          _key == key && _config == config
+        end
         conn.disconnect! if conn
         config
       end
```

**Comment by juhaerk** [comment]
Oh, the problem was encountered with the 5.1.0 version. I did not test the 5.2.0 version, but the patch applies to that version as well, thus I suspect the issue is also in 5.2.0.

**Comment by juhaerk** [comment]
After some testing I suspect my previous patch is bad. Some connections, that should be removed, do not get removed...

**Comment by kou** [comment]
Can you show a script that reproduces this case?

**Comment by kou** [comment]
Thanks.
It reports "All is as expected." without any changes. (I commented out `'method' => 'tls'` because my test LDAP server doesn't support TLS.)

The script may miss something to reproduce this problem.

**Comment by juhaerk** [comment]
In my environment, if the above script is run with the following wrapper (so that ActiveLDAP 5.2.0 is surely used), the issue triggers:

```
#!/bin/sh

set -eu

rm -rf .tmpgem

export GEM_HOME=.tmpgem

cat <<'EOF' > Gemfile
source 'https://rubygems.org'
gem 'activeldap', '5.2.0'
EOF

bundle install

bundle exec ruby activeldap_issue_test.rb
```

Here `activeldap_issue_test.rb` is the script I sent previously (or a modified version of it).

**Comment by kou** [comment]
Thanks but I couldn't reproduce this by it.

I commit a change that may fix this case. Can you try master?

**Comment by juhaerk** [comment]
Thank you, the code in master appears to work.

**Comment by kou** [comment]
Thanks for confirming it.
I've released 5.2.1.
